src/theory/strings/theory_strings_preprocess.cpp

```diff
@@ -251,6 +251,92 @@ Node StringsPreprocess::reduce(Node t,
     // Thus, indexof( x, y, n ) = skk.
     retNode = skk;
   }
+  else if (t.getKind() == kind::STRING_INDEXOF_RE)
+  {
+    // processing term:  indexof_re(s, r, n)
+    Node s = t[0];
+    Node r = t[1];
+    Node n = t[2];
+    Node skk = sc->mkTypedSkolemCached(
+        nm->integerType(), t, SkolemCache::SK_PURIFY, "iork");
+
+    Node negone = nm->mkConst(Rational(-1));
+    Node sLen = nm->mkNode(STRING_LENGTH, s);
+
+    // rem = substr(s, n, len(n) - n)
+    Node rem = nm->mkNode(STRING_SUBSTR, s, n, nm->mkNode(MINUS, sLen, n));
+    Node k1 =
+        sc->mkSkolemCached(rem, r, SkolemCache::SK_FIRST_MATCH_PRE, "iorpre");
+    Node k2 = sc->mkSkolemCached(rem, r, SkolemCache::SK_FIRST_MATCH, "iorm");
+    Node k3 =
+        sc->mkSkolemCached(rem, r, SkolemCache::SK_FIRST_MATCH_POST, "iorpost");
+
+    Node sigma = nm->mkNode(REGEXP_SIGMA);
+    Node sigmaStar = nm->mkNode(REGEXP_STAR, sigma);
+    Node anyMatchRe = nm->mkNode(REGEXP_CONCAT, sigmaStar, r, sigmaStar);
+    // ~in_re(rem, _* ++ r ++ _*)
```

Comment:
I don't think the reduction requires use of `_*`.  Doesnt the use of substr suffice?

Reply:
The problem is that we want to say that there is no match of `r` in `s`. For example, consider `str.indexof("abc", str.to_re("b"), 0)`. We want to check whether there is any match of the regular expression `str.to_re("b")` in `"abc"`. To do that, we check whether `"abc"` matches `re.++(_*, str.to_re("b"), _*)`

Reply:
There should be a better way to do this.  My concern here is that this condition is similar to the part that says there are no matches before the return index.

The idea would be that there is a quantified formula that says "there are no matches before k", where `k` is either 0 if `n` is out of bounds, `skk` if `skk` is not -1, and `len(s)` otherwise.